### DISCUSSION TITLE
On the use of dynamic vertex attribute buffers in WebGPU

### DISCUSSION THREAD

**Discussion by ghadeeras** [gpuweb/gpuweb#2209] [Q&A] [answered]
Hello!

Assuming I have not overlooked anything in the WebGPU specs, my understanding is that WebGPU buffers are not dynamic. That is, their sizes are determined at creation-time, and they cannot grow/shrink from one frame to the next. In contrast, WebGL "appears" to be able to resize the buffer upon calling `gl.bufferData(...)` with different amount of data.

I understand that maybe the majority of applications do not need such size volatility for their vertex attribute buffers, but I could think of cases where this might be desired. One example is when using the marching cubes algorithm to polygonise a scalar field isosurfaces. If the rendered isosurface changes from one frame to the next, so does the number of vertices (like in a classic meta-balls example).

When I first attempted to convert an example I wrote in the past (that used marching cubes), I first tried to recreate the buffer every time a resize is needed, simply by destroying it and immediately creating a new one with new vertex data which will get bound to the pipeline on the next render pass. That caused some weird behavior including the loss of the GPUDevice, which made me think that it could be one of the following:

   * My code has a silly bug that I failed to spot. 
   * The current WebGPU implementations in Chrome/Firefox are still buggy and I should have maybe reported this as a defect.
   * It is maybe unsafe to call the destroy method on a buffer that is potentially still in use by the GPU, and I should find a way to synchronize between the two timelines.
   * This is not the idiomatic way to do this in WebGPU. There are maybe alternative ways to do it.

Unlike reading/writing to buffers, the specs are not quite clear on how destroying a buffer plays out on the different timelines. Is it an asynchronous operation, and does it internally request the buffer to be returned to the host CPU before actually destroying the buffer? Or does it destroy it irrespective of any GPU usage?

In general, I wonder if there are guidelines or best practices about such type of usage. The only way it worked for me is with some awkward code where I juggle buffers between two variables. But I am not sure if this is working by some lucky coincidences. I also thought of estimating the maximum size of the buffer, so that I use one buffer with maximal size. But it is not very evident how to do that in the marching cubes example.

Sorry for the long and verbose question! Thanks in advance for any thoughts/answers.

**Comment by Kangz** [comment, marked as answer]
> Assuming I have not overlooked anything in the WebGPU specs, my understanding is that WebGPU buffers are not dynamic. That is, their sizes are determined at creation-time, and they cannot grow/shrink from one frame to the next. In contrast, WebGL "appears" to be able to resize the buffer upon calling gl.bufferData(...) with different amount of data.

That's correct. Internally the OpenGL driver will create a new buffer, copy data, etc for you. (see the sadness in [these slides](https://docs.google.com/presentation/d/12YSEcLjyYStNDK9EwhPgRq05oIxpTIqC1GdKgJd_Jt4/edit#slide=id.g1856010b44_0_1118)) More modern / explicit APIs require that if a resize needs to happen then it needs to be done by the applications.

> That caused some weird behavior including the loss of the GPUDevice, which made me think that it could be one of the following

Recreating vertex buffers many times shouldn't cause a `GPUDevice` loss and instead, at worst, produce an out of memory error. If you have a simple repro case can you report this on the respective browser bug tracker? Note that it is safe in WebGPU to destroy a buffer that's in flight, and it is the browser's responsibility to wait until the GPU is finished before deallocating it.

> In general, I wonder if there are guidelines or best practices about such type of usage. The only way it worked for me is with some awkward code where I juggle buffers between two variables. But I am not sure if this is working by some lucky coincidences.

I'm not sure I understand the two variable thing. What are you doing that works?

**Reply by ghadeeras** [reply]
Thanks a lot @Kangz !

What I meant by juggling buffers between two variables is a double buffering sort of thing. I have two variables A and B pointing to different GPU buffers. The piece of code that generates new vertex data (in an event handler) can only supply/recreate buffer A, while the animation frame rendering code consumes buffer B only (delaying its destruction until the following animation frame), and moves buffer A to B in preparation for the next frame.

I will try to write a minimal example that reproduces this issue before reporting it in the respective browser's bug tracker.

Indeed I expected that supporting dynamic buffers must have important disadvantages that led to dropping it from modern APIs. Thanks for sharing the presentation slides!

**Reply by Kangz** [reply]
I'm glad I could help. Looking forward to the issues, all easy to repro GPU losses are great so we can make implementations more robust.
